covariancematrix.go

```diff
@@ -38,35 +37,42 @@ func CovarianceMatrix(cov *mat64.Dense, x mat64.Matrix, wts []float64) *mat64.De
 	// Subtract the mean of each of the columns.
 	for i := 0; i < c; i++ {
 		v := xt.RawRowView(i)
+		// This will panic with ErrShape if len(wts) != len(v), so
+		// we don't have to check the size later.
 		mean := Mean(v, wts)
 		floats.AddConst(-mean, v)
 	}
 
-	var xc mat64.Dense
-	xc.TCopy(&xt)
-
 	var n float64
 	if wts != nil {
-		if wr := len(wts); wr != r {
-			panic(mat64.ErrShape)
-		}
 
+		// Multiply by the sqrt of the weights, so that multiplication is symmetric.
+		sqrtwts := make([]float64, r)
+		for i, w := range wts {
+			if w < 0 {
```

Comment:
It seems like the kind of thing we would panic on, but without the check it would produce a matrix of NaNs.  The nice thing about panic is that if we expand out to allow negative weights, callers to the function would (probably) not have to change anything.  If it was all NaN, they might have some code checking for it.  I don't feel very strongly about it though.